### PR TITLE
[5.1] Form: Add regex validation rule

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
This PR adds a regex validation rule for the Form library. The base class for Form validation rules has validation with regular expression implemented, but no way to hand in a regular expression via the form definition. You have to build your own validation rule class, register it in Joomla and can then use it. That seems overly complex for this, so this PR adds a rule to add regular expression validation.

For your field, you add `validate="regex"` to your form definition and the regular expression via `validate_regex=""`. The value of the field is then validated server side against the regex given. You have to pay attention that you might have to escape the regex properly when adding it to the validate_regex attribute. You can also add a modifier at the end.

### Testing Instructions
1. Create a new instance of a core module, for example mod_custom in the backend and go to the "Advanced" tab. Modify the value of one of the entries in the "Layout" select to be a value which does not contain A-Z, a number, - or _. Invalid would for example be `üüü`.
2. When saving, you should get an error, that this is invalid.
3. Edit /modules/mod_custom/mod_custom.xml and replace `validate="moduleLayout"` with 
	```
					validate="regex"
					validate_regex="^([A-Za-z0-9_-]+:)?[A-Za-z0-9-][A-Za-z0-9\.-]*$"
	```
4. Try the same as in 1. and see that it still fails, but that the other values still properly pass.

Or it could be approved by maintainers by codereview.

### Thanks
A big thanks goes to @coolcat-creations ([Website](https://coolcat-creations.com/)) who sponsored this PR.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [X] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/228
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by webnet-assmann** [comment]
Hello Hannes,
I've tested it, but I don't know if I understood it correctly:
You have to select something in the Advanced tab under Layout and enter a value with mutated vowel. I can't change the values there, so I created another file for Layout with üüü and then selected that. Then I got the error message. Then I changed the corresponding code in the mod_custom.xml file and selected üüü again and the error message appeared again, but the other values could be saved.
If I have understood correctly, then the test was successful.

**Comment by Hackwar** [comment]
Thanks @webnet-assmann. Yes, that would be a successfull test. The goal was to show that the regex validation rule properly functions, regardless if we use the specific rule for moduleLayout or if we copy the regex from there into the XML.

File: libraries/src/Form/Rule/RegexRule.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * @since  __DEPLOY_VERSION__
+ */
+class RegexRule extends FormRule
+{
+    /**
+     * Method to test the value.
+     *
+     * @param   \SimpleXMLElement  $element  The SimpleXMLElement object representing the `<field>` tag for the form field object.
+     * @param   mixed              $value    The form field value to validate.
+     * @param   string             $group    The field name group control value. This acts as as an array container for the field.
```
Comment (brianteeman) [inline libraries/src/Form/Rule/RegexRule.php]:
```suggestion
     * @param   string             $group    The field name group control value. This acts as an array container for the field.
```

**Comment by richard67** [comment]
> Tested successfully
> 
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/42657](https://issues.joomla.org/tracker/joomla-cms/42657).

@webnet-assmann Could you mark your test result in the issue tracker so it’s properly counted? Just go to https://issues.joomla.org/tracker/joomla-cms/42657 , use the blue „Test this“ button at the top left corner, then select your t st result and submit. Thanks in advance.

**Comment by brianteeman** [comment]
This really needs sopme documentation please

**Comment by coolcat-creations** [comment]
What would be the right place to add the documentation before the PR is merged?

**Comment by coolcat-creations** [comment]
For example:

- Email Filtering: Exclude specific domains (e.g., block freemail accounts). For example exclude gmail and web.de
- ZIP Code Validation: Ensure format consistency for different countries. For example 5 Numbers in Germany
- IBAN Validation: Validate international bank account numbers.
- Telephone Numbers: Accommodate country-specific formats and codes. For example +49 170 1234456 against (0170) 1234456
- Social Security Numbers: Validate format based on country-specific standards.
- Vehicle License Plates: Match specific national or regional patterns. 
- Currency Formatting: Ensure correct currency symbol and format. 
- Time Formats: Validate time in various international formats in a text field For example: 10.00–18.00 Uhr
- Postal Address Formatting: Check for country-specific postal address formats.

**Comment by brianteeman** [comment]
@coolcat-creations All of that is already possible using pattern (https://www.html5pattern.com/) or am I missing some big difference that requires this extra code

**Comment by Fedik** [comment]
@brianteeman html5 pattern is client side, what is in PR is server side.
It is probably possible to use the same pattern for both client ans server (and so simplify the field set up), but I did tested such thing, cannot say for sure.

File: libraries/src/Form/Rule/RegexRule.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     * @since   __DEPLOY_VERSION__
+     * @throws  \UnexpectedValueException if rule is invalid.
+     */
+    public function test(\SimpleXMLElement $element, $value, $group = null, Registry $input = null, Form $form = null)
+    {
+        if ((string) $element['validate_regex']) {
+            $this->regex = (string) $element['validate_regex'];
+        }
+
+        if ((string) $element['validate_modifier']) {
```
Comment (brianteeman) [inline libraries/src/Form/Rule/RegexRule.php:49]: What does this do?

**Comment by brianteeman** [comment]
agree when it is security related such as with a password but in these examples client side is much faster and provides a better user experience

File: libraries/src/Form/Rule/RegexRule.php (hunk 2 above)
Comment (richard67) [reply libraries/src/Form/Rule/RegexRule.php:49]: Modifier of a regex is the part after the closing of the pattern, e.g. in a regex like `/^[a-z]$/i` it is the `i` modifier for case-insensitivity.

Comment (brianteeman) [reply libraries/src/Form/Rule/RegexRule.php:49]: I asked as it was not mentioned in the PR code comments and wasnt obvious as the example regex  was case insensitive. Thanks for the explanation.

**Comment by wilsonge** [comment]
> agree when it is security related such as with a password but in these examples client side is much faster and provides a better user experience

We basically need both. Server side is always going to be required regardless (if nothing else we have an API still to maintain which can't do HTML5 validation). Many of the security issues we had in 3.9/3.10 were people intercepting requests from forms to the server and then making further modifications because we weren't doing proper server side validation (only client side). For regex things it's hard to say what the business implications might be - but I'm sure there's going to be some business cases things like phone numbers might be required.

TLDR: I think this is fine - combining the server and client side regex together into a single field if provided definitely a nice to have though.

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 70a3a536f61e0da389eba91ce0d93f9c5db11ea8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42657">issues.joomla.org/tracker/joomla-cms/42657</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42657">issues.joomla.org/tracker/joomla-cms/42657</a>.</sub>

**Comment by Hackwar** [comment]
I added documentation for all validation rules to the manual. Hope that is enough like this.
